storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -4017,122 +4014,1889 @@ dberr_t fil_space_t::garbage_collect(bool shutdown)
+                            m_old_xdes, m_old_xoffset, m_mtr);
+      fil_system.sys_space->free_len++;
+    }
+  }
+public:
+  PageOperator(buf_block_t *header_block, buf_block_t *iblock,
+               fseg_inode_t *inode,
+               uint32_t extent_size, byte* old_page_no,
+               mtr_t *mtr) :
+  m_header_block(header_block),
+  m_iblock(iblock), m_inode(inode), m_extent_size(extent_size),
+  m_mtr(mtr)
+  {
+    if (old_page_no)
+      memcpy(m_old_page_no, old_page_no, 4);
+    m_free_limit= mach_read_from_4(FSP_HEADER_OFFSET + FSP_FREE_LIMIT +
+                                   m_header_block->page.frame);
+    m_seg_id= mach_read_from_8(m_inode + FSEG_ID);
+  }
+
+  ~PageOperator()
+  {
+    for (buf_block_t *old : m_old_pages)
+      if (old)
+      {
+        old->page.hash= nullptr;
+        buf_block_free(old);
+      }
+  }
+
+
+  /** Get allocated new block */
+  buf_block_t* get_new_block() const noexcept { return m_new_block; }
+
+  /** Prepare the new page allocation from the new given extent
+  @param new_extent starting page of new extent
+  @param segment    segment allocation
+  @return error code */
+  dberr_t prepare_new_page(uint32_t new_extent, bool segment) noexcept
+  {
+    dberr_t err= DB_SUCCESS;
+    uint32_t size= mach_read_from_4(FSP_HEADER_OFFSET + FSP_SIZE
+                                    + m_header_block->page.frame);
+    if (new_extent >= size || new_extent >= m_free_limit)
+      return DB_CORRUPTION;
+
+    uint32_t new_descr_page_no= xdes_calc_descriptor_page(0, new_extent);
+    m_new_xdes= fsp_get_latched_page(page_id_t{0, new_descr_page_no},
+                                     m_mtr, &err);
+    if (!m_new_xdes)
+      return err;
+
+    ut_ad(!m_new_block);
+    m_ioffset= uint16_t(m_inode - m_iblock->page.frame);
+    m_need_segment= segment;
+    m_xoffset= uint16_t(xdes_calc_descriptor_index(0, new_extent) * XDES_SIZE
+                          + XDES_ARR_OFFSET + XDES_FLST_NODE);
+    m_new_descr= m_new_xdes->page.frame + m_xoffset - XDES_FLST_NODE;
+    m_new_state= uint32_t(xdes_get_state(m_new_descr));
+    uint32_t new_page= 0;
+
+    /* Allocate the new extent and initialize the extent state
+    with XDES_FSEG/XDES_FREE_FRAG */
+    if (m_new_state == XDES_FREE)
+    {
+      if (segment) err= initialize_segment_prepare();
+      else err= initialize_free_frag_prepare();
+
+      if (err) return err;
+new_page:
+      new_page= xdes_find_free(m_new_descr);
+      if (new_page == FIL_NULL)
+        return DB_CORRUPTION;
+
+      new_page+= new_extent;
+      buf_block_t *free_block= buf_LRU_get_free_block(have_no_mutex);
+      if (!free_block) return DB_OUT_OF_MEMORY;
+      m_new_block= buf_page_create(fil_system.sys_space, new_page, 0,
+                                   m_mtr, free_block);
+      if (m_new_block != free_block)
+        buf_pool.free_block(free_block);
+      fsp_init_file_page(fil_system.sys_space, m_new_block, m_mtr);
```

Comment:
If we used `have_no_mutex_soft`, we could run out of memory. But we aren’t. I think that it would be simpler to just invoke `fsp_page_create()` instead of duplicating its logic, and *not* check for `nullptr` return value, because that function would wait until a buffer page is available.